File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
        void fillGraph(TreeNode node,ArrayList<Integer>[] graph) {
        if(node==null) return;
        
        if(node.left!=null) {
             graph[node.val].add(node.left.val);
            graph[node.left.val].add(node.val);
        }
        if(node.right != null) {
            graph[node.val].add(node.right.val);
            graph[node.right.val].add(node.val);
        }
        
        fillGraph(node.left,graph);
        fillGraph(node.right,graph);
    }
    
    int bfs(ArrayList<Integer>[] graph,int start) {
        boolean[] vis = new boolean[100001];
        int ans = 0;
        LinkedList<Integer> que = new LinkedList<>();  
        que.addLast(start);
        
        while(que.size()!=0) {
            int size = que.size();
            while(size-->0) {
                int removeN = que.removeFirst();
                vis[removeN] = true;
                for(int i=0;i<graph[removeN].size();i++) {
                    if(!vis[graph[removeN].get(i)]) {
                        que.addLast(graph[removeN].get(i));
                    }
                }        
            }
                ans++;
        }
        return ans;
    }
        
        public int amountOfTime(TreeNode root, int start) {
            ArrayList<Integer>[] graph = new ArrayList[100001];
            for(int i=0;i<100001;i++) {
                graph[i] = new ArrayList<>();
            }
            fillGraph(root,graph);

            return bfs(graph,start)-1;
        }
}

File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void fillGraph(TreeNode* node,vector<vector<int>>& graph) {
        if(!node) return;
        
        if(node->left) {
             graph[node->val].push_back(node->left->val);
            graph[node->left->val].push_back(node->val);
        }
        if(node->right) {
            graph[node->val].push_back(node->right->val);
            graph[node->right->val].push_back(node->val);
        }
        
        fillGraph(node->left,graph);
        fillGraph(node->right,graph);
    }
    
    int bfs(vector<vector<int>>& graph,int start) {
        vector<bool> vis(100001);
        int ans = 0;
        queue<int> que ; que.push(start);
        while(que.size()!=0) {
            int size = que.size();
            while(size-->0) {
                int removeN = que.front(); que.pop();
                vis[removeN] = 1;
                for(int i=0;i<graph[removeN].size();i++) {
                    if(!vis[graph[removeN][i]]) {
                        que.push(graph[removeN][i]);
                    }
                }        
            }
                ans++;
        }
        return ans;
    }
    int amountOfTime(TreeNode* root, int start) {
        vector<vector<int>> graph(100001);
        fillGraph(root,graph);
        
        return bfs(graph,start)-1;
    }
};